cpp/ql/src/semmle/code/cpp/models/implementations/StdString.qll

```diff
@@ -296,23 +296,24 @@ class StdBasicOStream extends TemplateClass {
 }
 
 /**
- * The `std::ostream` function `operator<<` (defined as a member function).
+ * The `std::ostream` functions `operator<<` (defined as a member function),
+ * `put` and `write`.
  */
 class StdOStreamOut extends DataFlowFunction, TaintFunction {
-  StdOStreamOut() { this.hasQualifiedName("std", "basic_ostream", "operator<<") }
+  StdOStreamOut() { this.hasQualifiedName("std", "basic_ostream", ["operator<<", "put", "write"]) }
 
   override predicate hasDataFlow(FunctionInput input, FunctionOutput output) {
     // flow from qualifier to return value
-    input.isQualifierAddress() and
+    input.isQualifierObject() and
     output.isReturnValue()
```

Comment:
I don't agree with this change. The qualifier object is a `stringstream`, but the return value is a `stringstream &`.

Reply:
Yeah I think you're right.  These methods return a reference to their qualifier (`*this`), and I think I got confused by the reference (I think I'm supposed to always think of references as pointers for data flow).

That said we lose some flows if I change it back to `isQualifierAddress`.  I'm looking into this.

Reply:
I've fixed the lost flows with a change to `DataFlowUtil.qll`.  Please check that though - I'm not entirely sure how the dataflow library is supposed to deal with references.